artichoke-backend/src/extn/core/regexp/trampoline.rs

```diff
@@ -12,14 +12,21 @@ pub fn initialize(
 ) -> Result<Value, Error> {
     if let Ok(existing) = unsafe { Regexp::unbox_from_value(&mut into, interp) } {
         if existing.is_literal() {
-            // NOTE: In Ruby 3.0.0+, this branch should return a `FrozenError`.
-            return Err(SecurityError::with_message("can't modify literal regexp").into());
+            return Err(FrozenError::with_message("can't modify literal regexp").into());
         }
         return Err(TypeError::with_message("already initialized regexp").into());
     }
     let (options, encoding) = interp.try_convert_mut((options, encoding))?;
     let regexp = Regexp::initialize(interp, pattern, options, encoding)?;
-    Regexp::box_into_value(regexp, into, interp)
+    let result = Regexp::box_into_value(regexp, into, interp);
+    if let Some(options) = options {
+        if options.is_literal() {
+            let mut value = result?;
+            value.freeze(interp)?;
+            return Ok(value);
+        }
+    }
+    result
```

Comment:
can we short circuit return with `?` and use `matches!` here? It will remove a level of nesting with the `if`s.

```suggestion
    let mut value = Regexp::box_into_value(regexp, into, interp)?;
    if matches!(options, Some(options) if options.is_literal()) {
        value.freeze(interp)?;
    }
    Ok(value)
```

Reply:
Thank you for the advice. The suggested syntax is much more succinct.